network/packet_listener_proxy.go

```diff
@@ -0,0 +1,120 @@
+}
+
+type packetListenerRequestSender struct {
+	proxyConn        net.PacketConn
+	writeIdleTimeout time.Duration
+	writeIdleTimer   *time.Timer
+}
+
+// NewPacketProxyFromPacketListener creates a new [PacketProxy] that uses the existing [transport.PacketListener]
+// to create connections to a proxy. You can use this function if you already have an implementation of
+// [transport.PacketListener] and would like to inject it into one of the network stacks (for example,
+// network/lwip2transport) as UDP traffic handlers.
+func NewPacketProxyFromPacketListener(pl transport.PacketListener) (PacketProxy, error) {
+	if pl == nil {
+		return nil, errors.New("pl must not be nil")
+	}
+	return &packetListenerProxyAdapter{
+		listener:         pl,
+		writeIdleTimeout: 30 * time.Second,
+	}, nil
+}
+
+// NewSession implements [PacketProxy].NewSession function. It uses [transport.PacketListener].ListenPacket to create
+// a [net.PacketConn], and constructs a new [PacketRequestSender] that is based on this [net.PacketConn].
+func (proxy *packetListenerProxyAdapter) NewSession(respWriter PacketResponseReceiver) (PacketRequestSender, error) {
+	if respWriter == nil {
+		return nil, errors.New("respWriter must not be nil")
+	}
+	proxyConn, err := proxy.listener.ListenPacket(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	reqSender := &packetListenerRequestSender{
+		proxyConn:        proxyConn,
+		writeIdleTimeout: proxy.writeIdleTimeout,
+	}
+
+	// Terminate the session after timeout with no outgoing writes (deadline is refreshed by WriteTo)
+	reqSender.writeIdleTimer = time.AfterFunc(reqSender.writeIdleTimeout, func() { reqSender.Close() })
+
+	// Relay incoming UDP responses from the proxy asynchronously until EOF, session expiration or error
+	go func() {
+		defer respWriter.Close()
+
+		// Allocate buffer from slicepool, because `go build -gcflags="-m"` shows a local array will escape to heap
+		slice := packetBufferPool.LazySlice()
+		buf := slice.Acquire()
+		defer slice.Release()
+
+		for {
+			n, srcAddr, err := proxyConn.ReadFrom(buf)
+			if err != nil {
+				// Ignore some specific recoverable errors
+				if errors.Is(err, io.ErrShortBuffer) {
+					continue
+				}
+				return
+			}
+			if _, err := respWriter.WriteFrom(buf[:n], srcAddr); err != nil {
+				return
+			}
+		}
+	}()
+
+	return reqSender, nil
+}
+
+// WriteTo implements [PacketRequestSender].WriteTo function. It simply forwards the packet to the underlying
+// [net.PacketConn].WriteTo function.
+func (s *packetListenerRequestSender) WriteTo(p []byte, destination net.Addr) (int, error) {
+	s.writeIdleTimer.Reset(s.writeIdleTimeout)
+	return s.proxyConn.WriteTo(p, destination)
+}
+
+// Close implements [PacketRequestSender].Close function. It closes the underlying [net.PacketConn]. This will also
+// terminate the goroutine created in NewSession because s.conn.ReadFrom will return [io.EOF].
+func (s *packetListenerRequestSender) Close() error {
+	s.writeIdleTimer.Stop()
```

Comment:
Do you need to drain the channel, as described in https://pkg.go.dev/time#Timer.Stop?

Reply:
I don't think so, [`AfterFunc` doesn't have the `C` initialized](https://cs.opensource.google/go/go/+/refs/tags/go1.20.6:src/time/sleep.go;l=164-170), it will cause deadlock if you do so (check this in the playground: https://go.dev/play/p/RK_zbuhNUdQ)